src/java/voldemort/server/VoldemortServer.java

```diff
@@ -516,6 +521,22 @@ public void goOffline() {
 
     public void goOnline() {
         getMetadataStore().setOfflineState(false);
+        List<StorageEngine<ByteArray, byte[], byte[]>> storageEngines =
+                storageService.getStoreRepository().getStorageEnginesByClass(ReadOnlyStorageEngine.class);
+
+        if (storageEngines.size() > 0) {
+            logger.info("Will attempt to removeRemoteObsoleteState() for " + storageEngines.size() + " Read-Only stores.");
+            for (StorageEngine storageEngine: storageEngines) {
+                StoreVersionManager storeVersionManager =
+                        (StoreVersionManager) storageEngine.getCapability(StoreCapabilityType.DISABLE_STORE_VERSION);
+                try {
+                    storeVersionManager.removeRemoteObsoleteState();
+                } catch (Exception e) {
+                    logger.error("Failed to removeRemoteObsoleteState() for store: " + storageEngine.getName(), e);
+                }
+            }
+        }
```

Comment:
Also on looking at this again, it does store by store. For LSG1 with 70 stores, it takes about 60 seconds to run on each node. With 30 nodes, it takes 30 minutes to set the cluster to online.

Can we do a bulk operation of deleting a node-id ?